### ISSUE TITLE
t.p.log.FileLogObserver has problems with timezone names (mostly on windows).

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11330]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#630|
|Type|defect|
|Created|2004-06-03 08:00:00Z|

Attachments:

* [log.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c9/2c9baea38488b23d572875080939b4cb778835b8/9d2f6769de649a7e12acb9f3ea915d0bfaff58f8.diff) (1170 bytes) - added by ash on 2005-06-08 04:26:20Z - 
* [log.py.diff2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c9/2c9baea38488b23d572875080939b4cb778835b8/a281ef23c1906f15324c5e2eb8ea9a0c13d2eb3d.diff2) (2039 bytes) - added by ash on 2005-06-09 09:35:57Z - 
* [log.py.diff3](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c9/2c9baea38488b23d572875080939b4cb778835b8/669947e4a5f06fc9499471e42eac7ac280233511.diff3) (2640 bytes) - added by ash on 2005-06-09 17:48:55Z - 
* [log.diff4](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c9/2c9baea38488b23d572875080939b4cb778835b8/8ab24582cdce464649769eebfa83997c9a1fbcce.diff4) (7347 bytes) - added by jerub on 2006-08-20 22:22:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__630 630
type__defect defect
reporter__jknight jknight
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1086249600000000 1086249600000000
changetime__1156121758000000 1156121758000000
version__ 
owner__ 
cc__exarkun cc__jknight cc__ash
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
At first I was going to just suggest that the default log format should use "%z" instead of "%Z".

Most pressingly, because on WinXP there is a report that %Z returns "Central Europe Daylight Time". 
Also because timezone abbreviations suck and overlap in different parts of the world. "EDT" is not a 
global abbreviation, and means different things in different places. So really, "%Z" is sucky even on unix.

However, %z is not a standard strftime format. Amazingly enough, there is no standard strftime format 
character for showing time timezone like "-0500".

So I dunno what the right solution is. Perhaps FileLogObserver should not use strftime, because it is 
stinky, and instead use a python format string date format specifier like "%(year)d/%(month)02d/
%(day)02d %(hour)02d:%(minute)02d %(tzoffset)" which would be given an appropriately filled out 
dictionary.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="ash's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ash commented|
|-|-|

```
#!html
<pre>
I would add that for me (under Russian Windows XP) even %z doesn't work. It
doesn't differ from %Z. 
What is worse is that under Windows strftime returns timezone name in local
language (Russian in my case) and in default system encoding (cp1251). Notice,
no unicode! It means that I see complete rubbish instead of timezone name in
console with the log. It happens because console encoding is cp866 by default on
Russian Windows. [1]
For this reason I changed the title of this issue.

To summarize it all, we have the following problems:
1. Short timezone names can overlap.
2. Long timezone names are 
  2.1 ...hmm... too long.
  2.2 ... written in local language and local encoding under non-english Windows.

I like jknight's proposal regarding python format string. It is even better than
 string.Template (new to Python 2.4), because you can specify precisions inside
a string.

[1] You can fix the "rubbish instead of timezone" bug using conversions
according to locale and terminal encodings (see the patch attached). But the
timezone name will still be long of course.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="ash's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ash commented|
|-|-|

```
#!html
<pre>
I have looked through builtin datetime module documentation (introduced in
Python 2.3). According to it, datetime.datetime.strftime method is guarantied to
support %z. It works like jknight describes: "-0500", for example.
One caveat: one has to implement datetime.tzinfo subclass in order for timezone
name to show up using %z. Fortunatelly, in datetime documentation there is a
sample tzinfo subclass that gets timezone information from the operating system.

So I have prepaired a patch that makes FileLogObserver use datetime module. I
have also changed %Z to %z in FileLogObserver.timeFormat
See log.py.diff2 attachment.

Advantages:
1. Format of FileLogObserver.timeFormat stays the same (in case someone changes it).
2. jknight is happy because timezone information now shows as "-0500"
3. me is happy (see message below) :-)

Disadvantage:
Python 2.3 dependancy
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="ash's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ash commented|
|-|-|

```
#!html
<pre>
Now I've prepaired a patch (log.py.diff3) that merges two previous patches.
Summary of changes:
- FileLogObserver uses datetime module
- FileLogObserver.timeFormat contains '%z' instead of '%Z'
- re-encode string returned by strftime using locale.getpreferredencoding() and
encoding attribute of file object (for non-english versions of Windows) [1] 

[1] I put re-encoding under:
"""if sys.platform == 'win32' and '%Z' in self.timeFormat"""
It's not necessary. Remove if you don't like it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Ready for review in timezone-logging-630

I chose not to use datetime.strftime due to the amount of code required to define a working timezone.  I would rather not duplicate the logic to compute the timezone offset string, but doing so seems better than dragging a mountain of crap in to satisfy datetime's crappiness.

The %Z transcoding might be a valuable feature, but I think it's a separate issue so I've left it alone in this branch.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_6"></a>|@jerub commented|
|-|-|

I've attached a new diff which makes the tests pass on OS-X in Australia. Please test this branch for sanity in another timezone, especially another timezone with DST.

I've also noted that twisted.python.log has pyflakes warnings, please fix them:
```
$ pyflakes twisted/python/log.py
twisted/python/log.py:234: undefined name 'theLogPublisher'
twisted/python/log.py:385: undefined name '_oldshowwarning'
twisted/python/log.py:432: undefined name 'logfile'
twisted/python/log.py:468: undefined name 'defaultObserver'
```

Once you've integrated the new patch, fixed the pyflakes warnings, tested the branch on a few buildslaves in different timezones, please merge :)).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jerub|
|-|-|

pyflakes warnings are from the try/except NameError construct, so not much to be done about them.

I incorporated fixes from your patch, mainly:

  * use utcfromtimestamp
  * adjust mktime value by time.timezone
  * dropped day comparison from strftime assertion, since the day might change in some timezones.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub set owner to @exarkun|
|-|-|

Thankyou for the Ultimate Quality. Merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [17889]) Merge timezone-logging-630

Author: exarkun
Reviewer: jerub
Fixes [#11330](https://github.com/twisted/twisted/issues/11330)

This replaces usage in the default case of time.strftime by the log module
with a custom time formatting routine.  It also replaces the timezone name
with a numeric offset.

Providing a custom time format string is still supported and will cause the
code to revert to strftime.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_10"></a>|@jyknight commented|
|-|-|

Uhm, how is it that comparing day, hour, and minute can't be done because of timezone issues, but that it's okay to compare year and month? Won't it now just fail during a certain time period each month, instead of during a certain time period each day?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [17892]) Merge timezone-logging-630       

Author: exarkun       
Reviewer: jerub  
Fixes [#11330](https://github.com/twisted/twisted/issues/11330)

This replaces usage in the default case of time.strftime by the log module
with a custom time formatting routine.  It also replaces the timezone name
with a numeric offset.

Providing a custom time format string is still supported and will cause the
code to revert to strftime.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Replying to [jknight](#note_10):
> Uhm, how is it that comparing day, hour, and minute can't be done because of timezone issues, but that it's okay to compare year and month? Won't it now just fail during a certain time period each month, instead of during a certain time period each day?

It uses a fixed time which is more than a full day away from a month/year boundary.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
